File: ra-idp/src/authentication/ports/user-repository.ts
/**
 * Layer 3 — Application Logic（ポート定義）
 */

import type { User } from '../../spec-bindings/schemas'

export interface UserRepository {
  findBySub(sub: string): Promise<User | null>
  findByUsername(tenant_id: string, username: string): Promise<User | null>
  /**
   * email アドレス (小文字正規化済み) で User を解決する。複数 user が同じ
   * email を持つことは登録時に禁ずる前提で、最初の 1 件 / null を返す。
   * email を持たない user は対象外。詳細は ADR-030。
   */
  findByEmail(tenant_id: string, email: string): Promise<User | null>
  /**
   * 全ユーザーを created_at 昇順で返す。管理 API のユーザー一覧 (ADR-031) で使う。
   * deleted_at が立った user は含めない。
   */
  findAll(tenant_id: string): Promise<User[]>
  save(user: User): Promise<void>
}

File: ra-idp-go/internal/identitymanagement/ports/user_repository.go
package ports

import (
	"context"

	"ra-idp-go/internal/shared/spec"
)

// UserRepository は IdentityManagement が所有する User aggregate の永続化境界。
type UserRepository interface {
	// FindBySub は ADR-036 の tombstone (`deleted_at != null`) を除外する。
	// 既に削除された user を含めて引きたい場合は FindBySubIncludingDeleted を使う。
	FindBySub(ctx context.Context, sub string) (*spec.User, error)
	// FindBySubIncludingDeleted は tombstone を含めて user を引く。
	// DeleteUser use case の冪等判定や監査経路から呼ばれる。
	FindBySubIncludingDeleted(ctx context.Context, sub string) (*spec.User, error)
	FindByUsername(ctx context.Context, tenantID, username string) (*spec.User, error)
	FindByEmail(ctx context.Context, tenantID, email string) (*spec.User, error)
	FindAll(ctx context.Context, tenantID string) ([]*spec.User, error)
	Save(ctx context.Context, user *spec.User) error
}
